feat: #7 로그인/회원가입 기능 추가

wda067 opened the pull request:
### 📌 PR 타입(하나 이상의 PR 타입을 선택해주세요)
- 기능 추가

### 📌 관련 이슈
#7 

### ✨ 반영 브랜치
feat/7 -> develop

### 📝 변경 사항
- [x] Member와 SocialAccount 엔티티 분리 설계
- [x] 요청 JSON 역직렬화 예외 처리
- [x] 신규 회원 로그인 시 회원가입 필요 응답 반환
- [x] 기존 회원은 로그인 처리
- [x] 회원가입 완료 시 회원 및 소셜 계정 생성

### ➕ 추가 메모
- 로그인 시 JWT 반환은 구현 필요

### 🐛 테스트 결과 (테스트 결과가 있다면 넣어주세요)

#### 소셜 로그인 요청 (기존 회원인 경우)
```
POST http://localhost:8080/api/auth/login
Content-Type: application/json

{
  "socialProvider": "KAKAO",
  "socialId": "existing-social-id-123",
  "email": "existing@example.com"
}
```

#### 응답 결과
```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "signupRequired": false,
    "tokens": null
  }
}
```

#### 소셜 로그인 요청 (신규 회원인 경우)
```
POST http://localhost:8080/api/auth/login
Content-Type: application/json

{
  "socialProvider": "NAVER",
  "socialId": "new-social-id-999",
  "email": "newuser@example.com"
}
```

#### 응답 결과
```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "signupRequired": true,
    "tokens": null
  }
}
```

#### 회원가입 요청 (소셜 정보 + 유저 정보)
```
POST http://localhost:8080/api/auth/signup
Content-Type: application/json

{
  "login": {
    "socialProvider": "NAVER",
    "socialId": "new-social-id-999",
    "email": "newuser@example.com"
  },
  "nickname": "홍길동",
  "birthday": "1998-01-03",
  "gender": "MALE"
}
```

#### 응답 결과
```
HTTP/1.1 200 
Content-Type: application/json

{
  "code": "SUCCESS",
  "message": "요청이 성공적으로 처리되었습니다.",
  "data": {
    "signupRequired": false,
    "tokens": null
  }
}
```

wda067 committed:
feat: #7 로그인/회원가입 기능 추가

- Member와 SocialAccount 엔티티 분리 설계
- 요청 JSON 역직렬화 예외 처리
- 신규 회원 로그인 시 회원가입 필요 응답 반환
- 기존 회원은 로그인 처리
- 회원가입 완료 시 회원 및 소셜 계정 생성

roomie00 commented on src/main/java/pyws/swyp/member/entity/Member.java:
개인적으로 Entity와 DTO에 @AllArgsConstructor 사용은 필드 순서에 의존하다 보니 코드 변경 시 무결성 문제를 만들 수 있다는 이유로 지양하는 편입니다.

필드 순서를 변경하거나 추가할 때 기존 생성 코드가 컴파일 에러 없이 깨질 수 있다고 알고 있습니다. AllArgsConstructor 제거하고 필요한 Builder 패턴에 대해 하단에 추가하며 사용하는 것이 어떨까요...? Builder 패턴으로만 사용한다면 굳이 달아둬서 혹시 모를 위험을 감수할 필요가 없을 것 같습니다.

이와 관련해서 잘 정리해둔 블로그 글 첨부합니다 ☺️
https://velog.io/@skb0516/AllArgsConstructor%EB%A5%BC-%EC%A7%80%EC%96%91%ED%95%98%EB%8A%94-%EC%9D%B4%EC%9C%A0-%EA%B7%B8%EB%A6%AC%EA%B3%A0-Builder%EC%99%80-%ED%95%A8%EA%BB%98-%EC%93%B8-%EB%95%8C

wda067 replied:
```java
@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Member extends BaseEntity {

    ...

    @Builder
    public Member(String email, String nickname, LocalDate birthDate, Gender gender) {
        this.email = email;
        this.nickname = nickname;
        this.birthDate = birthDate;
        this.gender = gender;
    }
}
```

이런 식으로 말씀이시죠? 
`@AllArgsConstructor`를 편리성만 생각해서 사용했었어서 해당 사항은 생각해본 적이 없었네요!
전 이렇게 가도 괜찮습니다!

roomie00 replied:
넵 첨부하신 예시 맞습니다! 감사합니다

roomie00 commented on src/main/java/pyws/swyp/auth/dto/AuthResponse.java:
저는 전에 멘토님의 추천을 받고 작성의 편의성을 위해 dto를 class가 아닌 record로 정의하는편입니다. 다만 이렇게 하면 dto의 값 변경을 할 수 없어 간혹가다 불편함이 생기더라구요. 사실 dto 값 변경이 그렇게 일반적인 케이스는 아니라고 생각되지만 class와 record 각각의 장단점이 있는 만큼 명근님이 class를 더 선호하시고 굳이 record의 필요성이 크지 않다고 생각되시면 제 코드를 변경하겠습니다!! -> 제가 아직 로컬에서만 작업한거라 명근님이 굳이 record 써보고싶다 이런게 아니시면 제가 바꾸겠습니다 :)

아래 관련 문서 첨부합니다.☺️
https://velog.io/@devjoong/DTOClass-vs-RecordJava-14-%EC%B0%A8%EC%9D%B4

wda067 replied:
저도 dto를 변경할 일이 있을 까 싶긴 하지만..
record를 사용해 본 적이 없어 이번 기회에 사용해보겠습니다~

roomie00 replied:
아 혹시 오해하셨을까 덧붙이자면 record 사용 시 dto 변경이 어렵다는 말이었습니다!!

경험해보고 싶으시다면 그럼 저도 record로 하겠습니다 :) dto 변경이 진짜 간혹 필요하다면 그거 한정으로 class 사용하는 방식으로 가도 좋을 것 같아요

roomie00 reviewed (commented):
전체 코멘트를 정리하자면
1. @AllArgsConstructor 사용 여부 (Entity/Dto 에서)
2. Dto에 Record 타입 사용 여부
3. 디스코드에 공유드린 Exception 관련

위 내용들이 다 정해지고, 그에 맞게 수정되면 merge 해도 될 것 같습니다!!

wda067 committed:
refactor: #7 공통 구조 개선

- Request/Response DTO를 record로 전환
- CustomException 추상 클래스 제거 및 일반 클래스로 전환
- 엔티티 @AllArgsConstructor 제거
- 엔티티 생성자에 @Builder 적용

roomie00 reviewed (approved):
👍 확인했습니다!
